source/common/upstream/cluster_manager_impl.cc

```diff
@@ -206,16 +207,69 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::api::v2::Bootstrap& bootstra
         bootstrap.cluster_manager().upstream_bind_config().source_address());
   }
 
+  // Cluster loading happens in two phases: first all the primary clusters
+  // are loaded, and then all the secondary clusters are loaded. As it
+  // currently stands all non-EDS clusters are primary and only EDS
+  // clusters are secondary.
+  // This two phase loading is done for the same reason that we have
+  // a two phase cluster initialization sequence: secondary clusters
+  // might depend on primary clusters, and thus we need to load
+  // secondary cluster dependencies first.
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
-    loadCluster(cluster, false);
+    // First load all the primary clusters.
+    if (cluster.type() != envoy::api::v2::Cluster::EDS) {
+      loadCluster(cluster, false);
+    }
   }
 
-  // We can now potentially create the CDS API once the backing cluster exists.
-  if (bootstrap.dynamic_resources().has_cds_config()) {
-    cds_api_ = factory_.createCds(bootstrap.dynamic_resources().cds_config(), eds_config_, *this);
-    init_helper_.setCds(cds_api_.get());
-  } else {
-    init_helper_.setCds(nullptr);
+  for (const auto& cluster : bootstrap.static_resources().clusters()) {
+    // Now load all the secondary clusters.
+    if (cluster.type() == envoy::api::v2::Cluster::EDS) {
+      loadCluster(cluster, false);
+    }
+  }
+
+  // All the static clusters have been loaded. At this point we can check for the
+  // existence of the v1 sds backing cluster, and the ads backing cluster.
+  // TODO(htuch): Add support for multiple clusters, #1170.
+  const ClusterInfoMap loaded_clusters = clusters();
+  if (bootstrap.dynamic_resources().deprecated_v1().has_sds_config()) {
+    const auto& sds_config = bootstrap.dynamic_resources().deprecated_v1().sds_config();
+    switch (sds_config.config_source_specifier_case()) {
```

Comment:
This switch statement is almost identical to the one in the subscription code. Is there any way to put it into a utility? Perhaps a utility that just checks for errors? Then that could be called here as well as in the subscription code.